kubeconfig/src/Kubernetes/KubeConfig.hs

```diff
@@ -0,0 +1,184 @@
+
+toMap :: [NamedEntity a s] -> Map.Map Text a
+toMap = Map.fromList . fmap (\NamedEntity {..} -> (name, entity))
+
+data AuthInfo = AuthInfo
+  { clientCertificate     :: Maybe (FilePath)
+  , clientCertificateData :: Maybe (Text)
+  , clientKey             :: Maybe (FilePath)
+  , clientKeyData         :: Maybe (Text)
+  , token                 :: Maybe (Text)
+  , tokenFile             :: Maybe (FilePath)
+  , impersonate           :: Maybe (Text)
+  , impersonateGroups     :: Maybe ([Text])
+  , impersonateUserExtra  :: Maybe (Map.Map Text [Text])
+  , username              :: Maybe (Text)
+  , password              :: Maybe (Text)
+  , authProvider          :: Maybe (AuthProviderConfig)
+  } deriving (Eq, Generic, Show, Typeable)
+
+authInfoJSONOptions = camelToWithOverrides
+    '-'
+    ( Map.fromList
+        [ ("tokenFile"           , "tokenFile")
+        , ("impersonate"         , "as")
+        , ("impersonateGroups"   , "as-groups")
+        , ("impersonateUserExtra", "as-user-extra")
+        ]
+    )
+
+instance ToJSON AuthInfo where
+  toJSON = genericToJSON authInfoJSONOptions
+
+instance FromJSON AuthInfo where
+  parseJSON = genericParseJSON authInfoJSONOptions
+
+data Context = Context
+  { cluster   :: Text
+  , authInfo  :: Text
+  , namespace :: Maybe Text
+  } deriving (Eq, Generic, Show, Typeable)
+
+contextJSONOptions =
+    camelToWithOverrides '-' (Map.fromList [("authInfo", "user")])
+
+instance ToJSON Context where
+  toJSON = genericToJSON contextJSONOptions
+
+instance FromJSON Context where
+  parseJSON = genericParseJSON contextJSONOptions
+
+data AuthProviderConfig = AuthProviderConfig
+  { name   :: Text
+  , config :: Maybe (Map.Map Text Text)
+  } deriving (Eq, Generic, Show)
+
+instance ToJSON AuthProviderConfig where
+  toJSON = genericToJSON $ camelToWithOverrides '-' Map.empty
+
+instance FromJSON AuthProviderConfig where
+  parseJSON = genericParseJSON $ camelToWithOverrides '-' Map.empty
+
+-- |Returns the currently active context.
+getContext :: Config -> Either String Context
+getContext Config {..} =
+    let maybeContext = Map.lookup currentContext (toMap contexts)
+    in  case maybeContext of
+            Just ctx -> Right ctx
+            Nothing  -> Left ("No context named " <> T.unpack currentContext)
+
+-- |Returns the currently active user.
+getAuthInfo :: Config -> Either String (Text, AuthInfo)
+getAuthInfo cfg@Config {..} = do
+    Context {..} <- getContext cfg
+    let maybeAuth = Map.lookup authInfo (toMap authInfos)
+    case maybeAuth of
+        Just auth -> Right (authInfo, auth)
+        Nothing   -> Left ("No user named " <> T.unpack authInfo)
+
+-- |Returns the currently active cluster.
+getCluster :: Config -> Either String Cluster
+getCluster cfg@Config {..} = do
+    Context {..} <- getContext cfg
+    let maybeCluster = Map.lookup cluster (toMap clusters)
+    case maybeCluster of
+        Just cluster -> Right cluster
```

Comment:
You're shadowing the cluster field here that is being implicitly brought in to scope by the `{..}`, which makes the code confusing to read - it looks like you have a forward reference. Can you rename one of them to remove the shadowing?

Reply:
Done